numpy/core/src/multiarray/arrayfunction_override.c

```diff
@@ -329,6 +335,96 @@ array_implement_array_function(
 }
 
 
+/*
+ * Implements the __array_function__ protocol for a Python function, as described in
+ * in NEP-18. See numpy.core.overrides for a full docstring.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_array_function(
+    PyObject *NPY_UNUSED(dummy), PyObject *positional_args)
+{
+    PyObject *implementation, *public_api, *relevant_args, *args, *kwargs;
+
+    if (!PyArg_UnpackTuple(
+            positional_args, "implement_array_function", 5, 5,
+            &implementation, &public_api, &relevant_args, &args, &kwargs)) {
+        return NULL;
+    }
+
+    /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
+     * in downstream libraries.
+     */
+    if (kwargs != NULL && PyDict_Contains(kwargs, npy_ma_str_like)) {
+        PyDict_DelItem(kwargs, npy_ma_str_like);
+    }
+
+    PyObject *res = array_implement_array_function_internal(
+        public_api, relevant_args, args, kwargs);
+
+    if (res == Py_NotImplemented) {
+        return PyObject_Call(implementation, args, kwargs);
+    }
+    return res;
+}
+
+
+/*
+ * Implements the __array_function__ protocol for C array creation functions
+ * only. Added as an extension to NEP-18 in an effort to bring NEP-35 to
+ * life with minimal dispatch overhead.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_c_array_function_creation(
+    const char *function_name, PyObject *args, PyObject *kwargs)
+{
+    PyObject *relevant_args;
+    PyObject *numpy_module = NULL, *public_api = NULL;
+    PyObject *result = NULL;
+
+    if (kwargs == NULL) {
+        return Py_NotImplemented;
+    }
+
+    /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
+     * in downstream libraries. If that key isn't present, return NULL and
+     * let originating call to continue.
+     */
+    if (PyDict_Contains(kwargs, npy_ma_str_like)) {
+        relevant_args = PyTuple_Pack(1,
+                PyDict_GetItem(kwargs, npy_ma_str_like));
+        if (relevant_args == NULL) {
+            return NULL;
+        }
+        PyDict_DelItem(kwargs, npy_ma_str_like);
+    }
+    else {
+        return Py_NotImplemented;
+    }
+
+    numpy_module = PyImport_Import(npy_ma_str_numpy);
+    if (numpy_module != NULL) {
```

Comment:
If this returns NULL you need to crash hard, not keep going without numpy.

Reply:
I agree with you, but in my experience this kind of error checking is usually added even if unlikely to fail. Would you like me to get rid of it?

Reply:
Please keep the error checking, but make it propagate errors.

Failure to do so can result in segfaults instead of helped error messages.

Reply:
Yeah, just return NULL immediately, also removes a level of indentation :).

Reply:
Done in https://github.com/numpy/numpy/pull/16935/commits/59234cbfe2bb70715208ba5203f4aafd9be06206 .